infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -249,12 +249,19 @@ private void fillHole(CorfuPayloadMsg<FillHoleRequest> msg, ChannelHandlerContex
         }
     }
 
+    /**
+     * Perform a prefix
+     * Here the token is not used as is to perform the trim as the epoch at which the checkpoint was completed
+     * might be old. Hence, we use the msg epoch to perform the trim. This should be safe provided that the
+     * trim is performed only on the token provided by the CheckpointWriter which ensures that the checkpoint
+     * was persisted. Using any other address to perform a trim can cause data loss.
+     */
     @ServerHandler(type = CorfuMsgType.PREFIX_TRIM)
     private void prefixTrim(CorfuPayloadMsg<TrimRequest> msg, ChannelHandlerContext ctx,
                             IServerRouter r) {
         try {
             TrimRequest req = msg.getPayload();
-            batchWriter.prefixTrim(req.getAddress());
+            batchWriter.prefixTrim(req.getAddress(), msg.getEpoch());
```

Comment:
Just to prevent the confusion maybe consider a prefix method with a long address (from Token) and epoch from the message.

Reply:
I feel that will create more confusion. We have only one prefixTrim. The epoch should only be the message epoch and address - trim sequence number.